R/get_delay_estimate.R

```diff
@@ -40,61 +42,144 @@
   expectation <- rep_tri
+
   # Find the column to start filling in
   start_col <- which(colSums(is.na(rep_tri)) > 0)[1]
 
   # Only fill in reporting triangle if it is incomplete
   if (!is.na(start_col)) {
-    for (co in start_col:(n_delays)) {
-      block_top_left <- rep_tri[1:(n_dates - co + 1), 1:(co - 1), drop = FALSE]
-      block_top <- rep_tri[1:(n_dates - co + 1), co, drop = FALSE]
-      mult_factor[co - 1] <- sum(block_top) / max(sum(block_top_left), 1)
-      block_bottom_left <- expectation[(n_dates - co + 2):n_dates, 1:(co - 1),
-        drop = FALSE
-      ]
-      # We compute the expectation so that we can get the delay estimate
-      expectation[(n_dates - co + 2):n_dates, co] <- mult_factor[co - 1] *
-        rowSums(
-          block_bottom_left
-        )
+    for (co in start_col:n_delays) {
+      start_row <- which(is.na(rep_tri[, co]))[1]
+      # Extract relevant blocks of the triangle
+      block_top_left <- .extract_block_top_left(rep_tri, co, n_dates, start_row)
+      block_top <- .extract_block_top(rep_tri, co, n_dates, start_row)
+
+      # Calculate multiplication factor
+      mult_factor <- .calculate_mult_factor(block_top, block_top_left)
+
+      # Extract block bottom left
+      block_bottom_left <- .extract_block_bottom_left(
+        expectation,
+        co,
+        n_dates,
+        start_row
+      )
+
+      # Compute expectations for bottom right
+      expectation[start_row:n_dates, co] <- .compute_expectations(
+        mult_factor,
+        block_bottom_left
+      )
     }
   }
 
+  return(expectation)
+}
 
-  # Use the completed reporting matrix to get the point estimate of the
-  # delay distribution
-  pmf <- colSums(expectation) / sum(expectation)
-  return(pmf)
+#' Extract the top left block of the triangle
+#'
+#' @param rep_tri Reporting triangle
+#' @param co Current column
+#' @param n_dates Number of dates
+#' @param start_row Starting row
+#' @return Top left block of the triangle
+#' @noRd
+.extract_block_top_left <- function(rep_tri, co, n_dates, start_row) {
+  return(rep_tri[1:(start_row - 1), 1:(co - 1), drop = FALSE])
+}
+
+#' Extract the top block of the triangle
+#'
+#' @inheritParams .extract_block_top_left
+#' @return Top block of the triangle
+#' @noRd
+.extract_block_top <- function(rep_tri, co, n_dates, start_row) {
+  return(rep_tri[1:(start_row - 1), co, drop = FALSE])
+}
+
+#' Extract the bottom left block of the triangle
+#'
+#' @param expectation Expectation matrix
+#' @param co Current column
+#' @inheritParams .extract_block_top_left
+#' @return Bottom left block of the triangle
+#' @noRd
+.extract_block_bottom_left <- function(expectation, co, n_dates, start_row) {
+  return(expectation[start_row:n_dates, 1:(co - 1), drop = FALSE])
+}
+
+#' Calculate multiplication factor
+#'
+#' @param block_top Top block
+#' @param block_top_left Top left block
+#' @return Multiplication factor
+#' @noRd
+.calculate_mult_factor <- function(block_top, block_top_left) {
```

Comment:
I know having all of these small functions is maximally modular, but I do have the slightest preference, for the purpose of code readability, for this, 
`.compute_expectations()` and `.calculate_pmf()` to just be in the `.fill_triangle()` function

Reply:
It is also about testing and reuse (i.e we had several copy and pasted bits of code). If you only have a slight preference I would prefer to keep it as it is.

Reply:
I'm leaving this as is for the reasons given and closing but happy to discuss further in an issue.